### PR TITLE
[TS migration] Migrate 'TagPicker' component to TypeScript

### PR DESCRIPTION
**Pull request by fvlvte**
### Details
Migrated TagPicker component to TypeScript.

### Fixed Issues
$ https://github.com/Expensify/App/issues/31955
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/components/TagPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    tag: string;
+
+    /** Callback to submit the selected tag */
+    onSubmit: () => void;
+
+    /**
+     * Safe area insets required for reflecting the portion of the view,
+     * that is not covered by navigation bars, tab bars, toolbars, and other ancestor views.
+     */
+    insets: SafeAreaInsetProps;
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:44]: Use `EdgeInsets` type instead
```suggestion
    insets: EdgeInsets;
```

File: src/pages/SafeAreaInsetProps.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+type SafeAreaInsetProps = {
```
Comment (VickyStash) [inline src/pages/SafeAreaInsetProps.ts:1]: This type isn't necessary, use `EdgeInsets` instead. Also this file is located in pages for some reason, remove it please

File: src/libs/PolicyUtils.ts [hunk 3]
```diff
@@ -184,14 +184,13 @@ function getTagListName(policyTags: OnyxEntry<PolicyTags>) {
...
-function getTagList(policyTags: OnyxCollection<PolicyTags>, tagKey: string) {
+function getTagList(policyTags: OnyxCollection<PolicyTags>, tagKey: string): PolicyTags {
     if (Object.keys(policyTags ?? {})?.length === 0) {
         return {};
     }
 
     const policyTagKey = tagKey ?? Object.keys(policyTags ?? {})[0];
-
-    return policyTags?.[policyTagKey]?.tags ?? {};
+    return (policyTags?.[policyTagKey]?.tags as unknown as PolicyTags) ?? {};
```
Comment (VickyStash) [inline src/libs/PolicyUtils.ts:193]: [Update](https://github.com/Expensify/App/pull/34168/files#diff-4ec85ea72f25a41093d479be537473c4580bc7deb065882749f27e8eb0eb4f75) PolicyTag type instead
```suggestion
    return policyTags?.[policyTagKey]?.tags ?? {};
```

File: src/components/TagPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+type TagPickerOnyxProps = {
+    /** Collection of tags attached to a policy */
+    policyTags: OnyxCollection<PolicyTags>;
+
+    /** List of recently used tags */
+    policyRecentlyUsedTags: OnyxCollection<RecentlyUsedTags>;
+
+    /** Should show the selected option that is disabled? */
+    shouldShowDisabledAndSelectedOption?: boolean;
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:23]: It shouldn't be in OnyxProps, move it to `TagPickerProps` type

File: src/components/TagPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+type TagPickerProps = TagPickerOnyxProps & {
+    /** The policyID we are getting tags for */
+    // eslint-disable-next-line react/no-unused-prop-types
+    policyID: string;
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:29]: Let's add a comment that policyID is used in the withOnyx HOC

File: src/components/TagPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    onSubmit: () => void;
+
+    /**
+     * Safe area insets required for reflecting the portion of the view,
+     * that is not covered by navigation bars, tab bars, toolbars, and other ancestor views.
+     */
+    insets: SafeAreaInsetProps;
+};
+
+function TagPicker({selectedTag, tag, policyTags, policyRecentlyUsedTags, shouldShowDisabledAndSelectedOption, insets, onSubmit}: TagPickerProps) {
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:47]:
```suggestion
function TagPicker({selectedTag, tag, policyTags, policyRecentlyUsedTags, shouldShowDisabledAndSelectedOption = false, insets, onSubmit}: TagPickerProps) {
```

**Review by VickyStash — CHANGES_REQUESTED** [review]

File: src/components/TagPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            onSelectRow={onSubmit}
+        />
+    );
+}
+
+TagPicker.displayName = 'TagPicker';
+
+export default withOnyx<TagPickerProps, TagPickerOnyxProps>({
+    policyTags: {
+        key: ({policyID}) => `${ONYXKEYS.COLLECTION.POLICY_TAGS}${policyID}` as typeof ONYXKEYS.COLLECTION.POLICY_TAGS,
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:128]:
```suggestion
        key: ({policyID}) => `${ONYXKEYS.COLLECTION.POLICY_TAGS}${policyID}`,
```

File: src/components/TagPicker.tsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    },
+    policyRecentlyUsedTags: {
+        key: ({policyID}) => `${ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_TAGS}${policyID}` as typeof ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_TAGS,
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:131]:
```suggestion
        key: ({policyID}) => `${ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_TAGS}${policyID}`,
```

File: src/components/TagPicker.tsx [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import * as PolicyUtils from '@libs/PolicyUtils';
+import type SafeAreaInsetProps from '@pages/SafeAreaInsetProps';
+import CONST from '@src/CONST';
+import ONYXKEYS from '@src/ONYXKEYS';
+import type {PolicyTags, RecentlyUsedTags} from '@src/types/onyx';
+import OptionsSelector from './OptionsSelector';
+
+type TagPickerOnyxProps = {
+    /** Collection of tags attached to a policy */
+    policyTags: OnyxCollection<PolicyTags>;
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:17]: Since key takes only one item from collection
```suggestion
    policyTags: OnyxEntry<PolicyTags>;
```

File: src/components/TagPicker.tsx [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    /** List of recently used tags */
+    policyRecentlyUsedTags: OnyxCollection<RecentlyUsedTags>;
```
Comment (VickyStash) [inline src/components/TagPicker.tsx:20]: Same
```suggestion
    policyRecentlyUsedTags: OnyxEntry<RecentlyUsedTags>;
```

**Review by VickyStash — CHANGES_REQUESTED** [review]

File: src/components/TagPicker.tsx (hunk 7 above)
Comment (fabioh8010) [reply src/components/TagPicker.tsx:129]: You can move with this suggestion @fvlvte , it turns out that the problem we discussed before was because you was typing `policyTags` as `OnyxCollection`, but it should be `OnyxEntry` like @VickyStash suggested.

File: src/components/TagPicker.tsx (hunk 8 above)
Comment (fabioh8010) [reply src/components/TagPicker.tsx:132]: Same

File: src/components/TagPicker/index.js
Comment (fabioh8010) [inline src/components/TagPicker/index.js:1]: Why did you rename the file to `TagPicker.tsx`? Should be `index.tsx`

**Review by fabioh8010 — COMMENTED** [review]
@fvlvte Please resolve comments, merge conflicts and fix errors

File: src/components/TagPicker/index.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    shouldShowDisabledAndSelectedOption?: boolean;
+};
+
+function TagPicker({selectedTag, tag, policyTags, policyRecentlyUsedTags, shouldShowDisabledAndSelectedOption = false, insets, onSubmit}: TagPickerProps) {
+    const styles = useThemeStyles();
+    const StyleUtils = useStyleUtils();
+    const {translate} = useLocalize();
+    const [searchValue, setSearchValue] = useState('');
+
+    const policyTagList = policyTags?.[tag]?.tags;
```
Comment (fabioh8010) [inline src/components/TagPicker/index.tsx:53]: Why did you stop using `PolicyUtils.getTagList`? Before it was like this:

```ts
const policyTagList = PolicyUtils.getTagList(policyTags, tag);
```

Comment (fabioh8010) [reply src/components/TagPicker/index.tsx:53]: To be able to use it again, make the following change  in `PolicyUtils` file:

```ts
function getTagList(policyTags: OnyxEntry<PolicyTags>, tagKey: string) {
```

File: src/components/TagPicker/index.tsx [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                name: selectedTag,
+                enabled: true,
+                accountID: null,
+            },
+        ];
+    }, [selectedTag]);
+
+    const enabledTags = useMemo(() => {
+        if (!shouldShowDisabledAndSelectedOption) {
+            return Object.values(policyTagList ?? {});
```
Comment (fabioh8010) [inline src/components/TagPicker/index.tsx:75]:
```suggestion
            return policyTagList;
```

File: src/components/TagPicker/index.tsx [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+function TagPicker({selectedTag, tag, policyTags, policyRecentlyUsedTags, shouldShowDisabledAndSelectedOption = false, insets, onSubmit}: TagPickerProps) {
+    const styles = useThemeStyles();
+    const StyleUtils = useStyleUtils();
+    const {translate} = useLocalize();
+    const [searchValue, setSearchValue] = useState('');
+
+    const policyTagList = policyTags?.[tag]?.tags;
+    const policyTagsCount = Object.values(policyTagList ?? {}).filter((policyTag) => policyTag.enabled).length;
```
Comment (fabioh8010) [inline src/components/TagPicker/index.tsx:54]:
```suggestion
    const policyTagsCount = Object.values(policyTagList).filter((policyTag) => policyTag.enabled).length;
```

File: src/components/TagPicker/index.tsx [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        ];
+    }, [selectedTag]);
+
+    const enabledTags = useMemo(() => {
+        if (!shouldShowDisabledAndSelectedOption) {
+            return Object.values(policyTagList ?? {});
+        }
+        const selectedNames = selectedOptions.map((s) => s.name);
+
+        return [...selectedOptions, ...Object.values(policyTagList ?? {}).filter((policyTag) => policyTag.enabled && !selectedNames.includes(policyTag.name))];
```
Comment (fabioh8010) [inline src/components/TagPicker/index.tsx:79]:
```suggestion
        return [...selectedOptions, ...Object.values(policyTagList).filter((policyTag) => policyTag.enabled && !selectedNames.includes(policyTag.name))];
```

File: src/types/onyx/PolicyTag.ts [hunk 15]
```diff
@@ -8,6 +8,7 @@ type PolicyTag = {
     /** "General Ledger code" that corresponds to this tag in an accounting system. Similar to an ID. */
     // eslint-disable-next-line @typescript-eslint/naming-convention
     'GL Code': string;
+    tags: PolicyTags;
```
Comment (fabioh8010) [inline src/types/onyx/PolicyTag.ts:11]:
```suggestion

    tags: PolicyTags;
```

Please add a comment

File: src/components/TagPicker/index.tsx [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    const enabledTags = useMemo(() => {
+        if (!shouldShowDisabledAndSelectedOption) {
+            return policyTagList;
+        }
+        const selectedNames = selectedOptions.map((s) => s.name);
+
+        return [...selectedOptions, ...Object.values(policyTagList).filter((policyTag) => policyTag.enabled && !selectedNames.includes(policyTag.name))];
+    }, [selectedOptions, policyTagList, shouldShowDisabledAndSelectedOption]);
+
+    //
```
Comment (fabioh8010) [inline src/components/TagPicker/index.tsx:83]:
```suggestion
```

**Review by fabioh8010 — APPROVED** [review]
Just apply my suggestion and you can open to cross-review

**Review by filip-solecki — APPROVED** [review]
LGTM! Just add tests steps and videos/screenshots

**Comment by fabioh8010** [comment]
@fvlvte Please add tests steps and videos/screenshots 🙂

**Comment by AndrewGable** [comment]
What's the latest here?

**Comment by AndrewGable** [comment]
Seems like a duplicate as well: https://github.com/Expensify/App/pull/36580

**Comment by fabioh8010** [comment]
@AndrewGable 

@kubabutkiewicz took over this issue since Kacper left the project, we decided to continue the work on [this PR](https://github.com/Expensify/App/pull/36580).

**Comment by fabioh8010** [comment]
@fvlvte Could you close this one please? Thanks
